x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/EsField.java

```diff
@@ -77,18 +90,27 @@ private DataType readDataType(StreamInput in) throws IOException {
         return DataType.readFrom(name);
     }
 
+    public static <A extends EsField> A readFrom(StreamInput in) throws IOException {
+        return ((PlanStreamInput) in).readEsFieldWithCache();
+    }
+
     @Override
     public void writeTo(StreamOutput out) throws IOException {
+        if (((PlanStreamOutput) out).writeEsFieldCacheHeader(this)) {
+            writeContent(out);
+        }
+    }
+
+    protected void writeContent(StreamOutput out) throws IOException {
         out.writeString(name);
         esDataType.writeTo(out);
-        out.writeMap(properties, StreamOutput::writeNamedWriteable);
+        out.writeMap(properties, (o, x) -> x.writeTo(out));
         out.writeBoolean(aggregatable);
         out.writeBoolean(isAlias);
     }
 
-    @Override
     public String getWriteableName() {
-        return ENTRY.name;
+        return getClass().getSimpleName();
```

Comment:
I'd prefer if the subclasses overrode this with a hard coded string. That way it's harder to break with a refactoring.